### PR TITLE
Fold bank serialisation into serde snapshot

### PR DESCRIPTION
**Pull request by svenski123**
#### Problem
Bank serialisation logic is currently in ledger/src/snapshot_utils and is not versioned

#### Summary of Changes
- Remove serde traits from Bank
- Move Bank related serde logic from snapshot_utils to serde_snapshot
- Add bank fields types representing format independent Bank persistent data
- Add crate visible ctor and field extract Bank for use by serde_snapshot

#### Notes
- This PR is dependent upon #10580
- Legacy and future snapshot formats are currently unchanged
- Future PR will remove bank fields not needed for serialisation from new future format (i.e. because fields are unused, fields can be recovered from genesis config, etc.)

### CODE REVIEW

File: runtime/src/bank.rs [hunk 1]
```diff
@@ -498,6 +547,95 @@ impl Bank {
...
 
+    /// Create a bank from explicit arguments and deserialized fields from snapshot
+    pub(crate) fn new_from_fields(
+        bank_rc: BankRc,
+        genesis_config: &GenesisConfig,
+        fields: BankFields,
+    ) -> Self {
+        let mut bank = Self {
+            rc: bank_rc,
+            src: Default::default(),
```
Comment (ryoqun) [inline runtime/src/bank.rs:558]: @svenski123 Well, this causes our custom sanity CI check failure: https://buildkite.com/solana-labs/solana/builds/25939#1ea7fd6a-912d-4911-8a57-046abaf2a617/78-79. Also, there are several similar ones.

Could you explicitly spell-out the types?

Comment (svenski123) [reply runtime/src/bank.rs]: I've reviewed #2630 / #2711 and the motivation behind avoiding the use of Default::default() in function calls to improve code readability.

In the struct initialization case here, the member variable names are present and the code readability concerns do not apply in this instance, I have therefore relied on Rust's type inference but in a way that does not violate the CI check.

**Comment by ryoqun** [comment]
@svenski123 As always, super thanks for working on this! It seems that you're really interested in this area. In that case, please be aware of my consistently-postponed attempt of establishing trust chain of snapshot https://github.com/solana-labs/solana/pull/8185  I think this pr should shed lights on the ultimate milestone of fully-verifiable snapshot. :)

Also, there is random notes for snapshot security in this messy meta-tracking issue, too #7167 if interested.

**Comment by svenski123** [comment]
Rebased and dependency on #10575 removed

**Comment by svenski123** [comment]
It's not clear to me why the buildkite buid #26406 job failed - solana_cli_config::Config::import_address_labels is clearly defined in cli-config/src/config.rs and this all builds fine locally.

I will rebase to master and force push again.

File: runtime/src/bank.rs [hunk 2]
```diff
@@ -498,6 +547,98 @@ impl Bank {
         new
     }
 
+    /// Create a bank from explicit arguments and deserialized fields from snapshot
+    pub(crate) fn new_from_fields(
+        bank_rc: BankRc,
+        genesis_config: &GenesisConfig,
+        fields: BankFields,
+    ) -> Self {
+        fn new<T: Default>() -> T {
```
Comment (ryoqun) [inline runtime/src/bank.rs:556]: Heh, I guess this acceptable evacuation. But `new` is so similar to the tranditional `::new()`. The `::new()` can take some arguments to initialize a new instance. At first glance, I'm confused with them.

How about renaming this to something like `default_generic()` or just `default()`?

File: runtime/src/bank.rs [hunk 3]
```diff
@@ -188,21 +187,81 @@ impl HashAgeKind {
...
+    pub(crate) collected_rent: u64,
+    pub(crate) rent_collector: RentCollector,
+    pub(crate) epoch_schedule: EpochSchedule,
+    pub(crate) inflation: Inflation,
+    pub(crate) stakes: Stakes,
+    pub(crate) epoch_stakes: HashMap<Epoch, EpochStakes>,
+    pub(crate) is_delta: bool,
+}
+
+pub(crate) struct RefBankFields<'a> {
```
Comment (ryoqun) [inline runtime/src/bank.rs:225]: Could you comment why `RefBankFields` is needed along side with just `BankFields`?

File: runtime/src/serde_snapshot/common.rs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+#[derive(Default, Clone, PartialEq, Debug, Deserialize, Serialize)]
+pub(crate) struct UnusedAccounts {
+    unused1: HashSet<Pubkey>,
+    unused2: HashSet<Pubkey>,
+    unused3: HashMap<Pubkey, u64>,
+}
+
+// Deserializable version of Bank for snapshot format
+#[derive(Clone, Deserialize)]
+pub(crate) struct DeserializableBankFields {
```
Comment (ryoqun) [inline runtime/src/serde_snapshot/common.rs:15]: Maybe `DeserializableBankFields` and `SerializableBankFields` cannot be unified (maybe unification introduces extra `.clone()`s at periodic snapshot creation)? If so, please add a rationale of it in comment in source code?

File: runtime/src/bank.rs (hunk 2 above)
Comment (svenski123) [reply runtime/src/bank.rs:556]: I wanted something short that sounded like a ctor.
"new" is three letters whereas "default"is more than twice that.
But fine to go with the latter if it's clearer.

**Comment by ryoqun** [comment]
I started to review this in more detail.

**Comment by ryoqun** [comment]
@svenski123 Like before, I've made some commits on top of your PR's topic branch: https://github.com/ryoqun/solana/commits/fold-bank-serialisation-into-serde-snapshot-01

Could you take a look at it? First of all, as always, I sensed that your pr is well thought-out. What I've done there wes rather cosmetic and to adjust our further direction a bit.

Before diving into the explanation of my branch, our pr's goal is...:

Make **versioning `Bank`'s (de-)serialization VERY simple even for other engineers not familiar with rather complex snapshot serialization mechanism. Make it possible for engineers to modify the Future (de-)serialization code as they wish until branching next version.**

With that in mind,

My branch contains 2 commits:
- https://github.com/ryoqun/solana/commit/4a3b5159c22cfe96c92b12d018bf6fcaeb1614d2: various renaming to clarify separated serialization and deserialization and to reflect structural symmetry for the two processings. Also, intentional code duplication to realize the mentioned goal. Intended to be chrry-picked by you to this pr. Also, I've added some explanations for `struct`s. 
- https://github.com/ryoqun/solana/commit/cd83759b04f835f589de889f1a1e4b420ce5a1b2: Not intended to be cherry-picked by you. Illustration-only. This is real demand for minor Bank serialization changes. As you guessed, `unused_accounts` are really unused now and `operating_mode` was added after our mainnet-beta launch and should generally be available in banks for convenience. Note that required changes are so small and easy, thanks to your work to this day! :)

File: runtime/src/bank.rs (hunk 2 above)
Comment (ryoqun) [reply runtime/src/bank.rs:556]: Ok, fine as-is! Come to think of it, `HashMap::new()` is completely legal in rust as well. I've updated my mental model. :)

File: runtime/src/bank.rs (hunk 3 above)
Comment (ryoqun) [reply runtime/src/bank.rs:225]: Commented by my self https://github.com/solana-labs/solana/pull/10581#issuecomment-652905768

File: runtime/src/serde_snapshot/common.rs (hunk 4 above)
Comment (ryoqun) [reply runtime/src/serde_snapshot/common.rs:15]: Commented by my self: https://github.com/solana-labs/solana/pull/10581#issuecomment-652905768

**Comment by svenski123** [comment]
Rebased to master & applied your requested patch, also worked in the abi stuff which is now in master.
AbiExample does not work with reference fields hence the workarounds.

I've aded assertions for bank fields which come from the genesis config, the intention is that these fields are removed from the snapshot entirely and the genesis config used instead.

File: runtime/src/bank.rs [hunk 5]
```diff
@@ -514,6 +548,136 @@ impl Bank {
...
+            bank.hashes_per_tick,
+            genesis_config.poh_config.hashes_per_tick
+        );
+        assert_eq!(bank.ticks_per_slot, genesis_config.ticks_per_slot);
+        assert_eq!(
+            bank.ns_per_slot,
+            genesis_config.poh_config.target_tick_duration.as_nanos()
+                * genesis_config.ticks_per_slot as u128
+        );
+        assert_eq!(bank.genesis_creation_time, genesis_config.creation_time);
```
Comment (ryoqun) [inline runtime/src/bank.rs:618]: @svenski123 (Thanks for continued work on this!) As a quick reminder, it seems that CI failed at this assertion:

https://buildkite.com/solana-labs/solana/builds/27132#3b380563-b5ce-418e-bb89-7a9012d83ae9/79-2158
```
failures:
    tests::V1_1_0::test_bank_forks_snapshot_n
    tests::V1_1_0::test_bank_forks_status_cache_snapshot_n
    tests::V1_2_0::test_bank_forks_snapshot_n
    tests::V1_2_0::test_bank_forks_status_cache_snapshot_n
```

Comment (svenski123) [reply runtime/src/bank.rs:618]: Yes it did fail, due to a bug in the bank_forks tests now fixed.

**Comment by ryoqun** [comment]
@svenski123 Cool, CI has passed! I'll review this in depth today or tomorrow. Also, now it seems there's a merge conflict...

**Comment by ryoqun** [comment]
@svenski123 It looks like CI currently fails to compile on rustc nightly.

**Comment by svenski123** [comment]
bincode upgrade in 841ecfd broke it; now fixed.
squashed the branch and applied the renamings including the code replication.

FYI the serdes bank fields were intentionally shared between legacy and future as both those formats are "live".
Rather future was to be renamed current and then a new future snapshot format added with its own (rather smaller) serdes bank fields def in a separate PR.

Of course that was three weeks ago.

**Comment by svenski123** [comment]
Ugh the renames broke the ABI hashes...

This is beginning to feel more like a hardware design process rather than a software design process...

**Comment by ryoqun** [comment]
@svenski123 I know the tedious feeling... However, the new abi tests were introduced to guard against accidental unintended abi changes. So, it's kind of expected to be annoying if you're intentionally modifying the abi-related code...

Previously, we're really bothered to do manual abi compatibility check every time, even taking into account of deeply-nested `struct`s. I hope this speeds up our development in overall.

**Pull request merged**